bbq/vm/test/vm_test.go

```diff
@@ -5563,3 +5563,133 @@ func TestUnclosedUpvalueAssignment(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, interpreter.NewUnmeteredIntValueFromInt64(3), actual)
 }
+
+func TestClosedUpvalue(t *testing.T) {
+
+	t.Parallel()
+
+	actual, err := compileAndInvoke(t,
+		`
+          fun new(): fun(Int): Int {
+              let x = 1
+              fun addToX(_ y: Int): Int {
+                  return x + y
+              }
+              return addToX
+          }
+
+          fun test(): Int {
+              let f = new()
+              return f(1) + f(2)
+          }
+        `,
+		"test",
+	)
+	require.NoError(t, err)
+	assert.Equal(t, interpreter.NewUnmeteredIntValueFromInt64(5), actual)
+}
+
+func TestClosedUpvalueAssignment(t *testing.T) {
```

Comment:
Not related to this test-case itself; but in the counterpart test `TestUnclosedUpvalueAssignment`, could you maybe add another variant where it calls the `addToX` more than once before returning?
e.g:
```cadence
fun test(): Int {
    var x = 1
    fun addToX(_ y: Int) {
        x = x + y
    }
    addToX(2)
    addToX(2) // call it another time: this time it should pick the updated `x`
    return x
}
```

Reply:
Good idea! Added in [eac8f6c](https://github.com/onflow/cadence/pull/3846/commits/eac8f6ce8aef3f5265984e3bb11c2d8140a27981)